File: test/fixtures/test-app.js
#!/usr/bin/env node

// Test Node.js application for PM2go E2E testing
console.log('Test Node.js app starting...');
console.log('Environment:', process.env.NODE_ENV || 'development');
console.log('Port:', process.env.PORT || 3000);
console.log('PID:', process.pid);

let counter = 0;

// Log every 5 seconds to generate log output for testing
setInterval(() => {
    counter++;
    console.log(`[${new Date().toISOString()}] Heartbeat #${counter} - PID: ${process.pid}`);
    
    // Test environment variables
    if (process.env.TEST_VAR) {
        console.log(`TEST_VAR: ${process.env.TEST_VAR}`);
    }
    
    // Simulate some work
    const memUsage = process.memoryUsage();
    console.log(`Memory: RSS=${Math.round(memUsage.rss/1024/1024)}MB, Heap=${Math.round(memUsage.heapUsed/1024/1024)}MB`);
    
}, 5000);

// Handle graceful shutdown
process.on('SIGTERM', () => {
    console.log('Received SIGTERM, shutting down gracefully...');
    process.exit(0);
});

process.on('SIGINT', () => {
    console.log('Received SIGINT, shutting down gracefully...');
    process.exit(0);
});

// Keep the process running
console.log('Test Node.js app is running. Use Ctrl+C to stop.');
console.log('Environment variables:');
Object.keys(process.env).forEach(key => {
    if (key.startsWith('TEST_') || key.startsWith('NODE_')) {
        console.log(`  ${key}=${process.env[key]}`);
    }
});

// Simulate crash scenario for testing restart functionality
if (process.env.CRASH_AFTER) {
    const crashTime = parseInt(process.env.CRASH_AFTER) * 1000;
    setTimeout(() => {
        console.log('Simulating crash for testing...');
        process.exit(1);
    }, crashTime);
}

File: test/fixtures/test-app.py
#!/usr/bin/env python3
"""
Simple test application for PM2go testing.
Produces output every 2 seconds with timestamps and optional arguments.
"""

import sys
import time
import argparse
import os
from datetime import datetime

def main():
    parser = argparse.ArgumentParser(description='PM2go test application')
    parser.add_argument('--interval', '-i', type=int, default=2, 
                       help='Output interval in seconds (default: 2)')
    parser.add_argument('--max-count', '-c', type=int, default=0,
                       help='Maximum number of outputs (0 = infinite)')
    parser.add_argument('--error-every', '-e', type=int, default=0,
                       help='Print to stderr every N iterations (0 = never)')
    parser.add_argument('--message', '-m', type=str, default='Hello from PM2go test app',
                       help='Custom message to output')
    parser.add_argument('--env-vars', action='store_true',
                       help='Print environment variables on startup')
    
    args = parser.parse_args()
    
    # Print startup info
    print(f"=== PM2go Test App Started ===")
    print(f"PID: {os.getpid()}")
    print(f"Args: {sys.argv}")
    print(f"Interval: {args.interval}s")
    print(f"Max count: {args.max_count if args.max_count > 0 else 'infinite'}")
    
    # Print environment variables if requested
    if args.env_vars:
        print("=== Environment Variables ===")
        for key, value in sorted(os.environ.items()):
            if key.startswith(('TEST_', 'PM2GO_', 'NODE_', 'PYTHON_')):
                print(f"{key}={value}")
        print("=============================")
    
    count = 0
    
    try:
        while True:
            count += 1
            timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            
            # Regular output to stdout
            print(f"[{timestamp}] #{count}: {args.message}")
            sys.stdout.flush()
            
            # Optional error output
            if args.error_every > 0 and count % args.error_every == 0:
                print(f"[{timestamp}] ERROR #{count}: This is an error message", file=sys.stderr)
                sys.stderr.flush()
            
            # Check if we've reached max count
            if args.max_count > 0 and count >= args.max_count:
                print(f"[{timestamp}] Reached max count ({args.max_count}), exiting")
                break
                
            time.sleep(args.interval)
            
    except KeyboardInterrupt:
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Received SIGINT, shutting down gracefully")
        print(f"Total outputs: {count}")
    except Exception as e:
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ERROR: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()